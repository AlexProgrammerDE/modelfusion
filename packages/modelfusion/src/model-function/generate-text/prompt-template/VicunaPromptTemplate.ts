import { TextGenerationPromptTemplate } from "../TextGenerationPromptTemplate.js";
import { ChatPrompt } from "./ChatPrompt.js";
import { validateContentIsString } from "./ContentPart.js";
import { InvalidPromptError } from "./InvalidPromptError.js";

// default Vicuna 1 system message
const DEFAULT_SYSTEM_MESSAGE =
  "A chat between a curious user and an artificial intelligence assistant. " +
  "The assistant gives helpful, detailed, and polite answers to the user's questions.";

/**
 * Formats a chat prompt as a Vicuna prompt.
 *
 * Overriding the system message in the first chat message can affect model responses.
 *
 * Vicuna prompt template:
 * ```
 * A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.
 *
 * USER: {prompt}
 * ASSISTANT:
 * ```
 */
export function chat(): TextGenerationPromptTemplate<ChatPrompt, string> {
  return {
    format(prompt) {
      let text =
        prompt.system != null
          ? `${prompt.system}\n\n`
          : `${DEFAULT_SYSTEM_MESSAGE}\n\n`;

      for (const { role, content } of prompt.messages) {
        switch (role) {
          case "user": {
            const textContent = validateContentIsString(content, prompt);
            text += `USER: ${textContent}\n`;
            break;
          }
          case "assistant": {
            text += `ASSISTANT: ${validateContentIsString(content, prompt)}\n`;
            break;
          }
          case "tool": {
            throw new InvalidPromptError(
              "Tool messages are not supported.",
              prompt
            );
          }
          default: {
            const _exhaustiveCheck: never = role;
            throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
          }
        }
      }

      // AI message prefix:
      text += `ASSISTANT: `;

      return text;
    },
    stopSequences: [`\nUSER:`],
  };
}
