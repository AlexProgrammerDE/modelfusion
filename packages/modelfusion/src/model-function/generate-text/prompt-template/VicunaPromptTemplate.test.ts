import { chat } from "./VicunaPromptTemplate.js";

describe("chat prompt", () => {
  it("should format prompt with user message", () => {
    const prompt = chat().format({
      messages: [{ role: "user", content: "user message" }],
    });

    expect(prompt).toMatchSnapshot();
  });

  it("should format prompt with user-assistant-user messages", () => {
    const prompt = chat().format({
      messages: [
        { role: "user", content: "1st user message" },
        { role: "assistant", content: "assistant message" },
        { role: "user", content: "2nd user message" },
      ],
    });

    expect(prompt).toMatchSnapshot();
  });

  it("should format prompt with system message and user-assistant-user messages", () => {
    const prompt = chat().format({
      system: "you are a chatbot",
      messages: [
        { role: "user", content: "1st user message" },
        { role: "assistant", content: "assistant message" },
        { role: "user", content: "2nd user message" },
      ],
    });

    expect(prompt).toMatchSnapshot();
  });
});
