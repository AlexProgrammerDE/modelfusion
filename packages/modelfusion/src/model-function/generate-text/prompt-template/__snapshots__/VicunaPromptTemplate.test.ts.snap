// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chat prompt > should format prompt with system message and user-assistant-user messages 1`] = `
"you are a chatbot

USER: 1st user message
ASSISTANT: assistant message
USER: 2nd user message
ASSISTANT: "
`;

exports[`chat prompt > should format prompt with user message 1`] = `
"A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.

USER: user message
ASSISTANT: "
`;

exports[`chat prompt > should format prompt with user-assistant-user messages 1`] = `
"A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.

USER: 1st user message
ASSISTANT: assistant message
USER: 2nd user message
ASSISTANT: "
`;
